### PR TITLE
Avoid DivisionByZeroError

### LINKED ISSUE #16463: DivisionByZeroError on sensors
### The problem

In a very small, DEV env, it happen with checkpoint devices. Not sure at this stage of only checkpoint are affected.
```
Error discovering sensors module for X.X.X.X. DivisionByZeroError: Division by zero in /opt/librenms/includes/discovery/sensors/entity-sensor.inc.php:152
Stack trace:
#0 /opt/librenms/includes/discovery/sensors.inc.php(14): require()
#1 /opt/librenms/includes/discovery/functions.inc.php(183): include('...')
#2 /opt/librenms/discovery.php(108): discover_device()
#3 {main}
```

### Output of ./validate.php

```txt
librenms@localhost:~$ ./validate.php
===========================================
Component | Version
--------- | -------
LibreNMS  | 24.9.0 (2024-09-30T10:12:12+01:00)
DB Schema | 2024_08_12_232009_ent_physical_table_rev_length (300)
PHP       | 8.3.11
Python    | 3.9.2
Database  | MariaDB 10.11.8-MariaDB-ubu2204
RRDTool   | 1.9.0
SNMP      | 5.9
===========================================

[OK]    Installed from package; no Composer required
[OK]    Database connection successful
[OK]    Database connection successful
[OK]    Database Schema is current
[OK]    SQL Server meets minimum requirements
[OK]    lower_case_table_names is enabled
[OK]    MySQL engine is optimal
[OK]    Database and column collations are correct
[OK]    Database schema correct
[OK]    MySQL and PHP time match
[OK]    Active pollers found
[OK]    Dispatcher Service is enabled
[OK]    Locks are functional
[OK]    Python wrapper cron entry is not present
[WARN]  Using database for locking, you should set CACHE_DRIVER=redis
[OK]    rrd_dir is writable
[OK]    rrdtool version ok
[WARN]  Non-git install, updates are manual or from package
```

### What was the last working version of LibreNMS?

Unknown

### Anything in the logs that might be useful for us?

All the $entry's that generated the 0.
```
{"entPhySensorType":"voltsDC","entPhySensorScale":"zepto","entPhySensorPrecision":"0","entPhySensorValue":"4979","entPhySensorOperStatus":"ok"}

{"entPhySensorType":"voltsDC","entPhySensorScale":"zepto","entPhySensorPrecision":"0","entPhySensorValue":"1794","entPhySensorOperStatus":"ok"}

{"entPhySensorType":"voltsDC","entPhySensorScale":"zepto","entPhySensorPrecision":"0","entPhySensorValue":"3295","entPhySensorOperStatus":"ok"}
```

### PR DESCRIPTION
**Pull request by Calvario**
Please give a short description what your pull request is for

Fix #16463 

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [X] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [ ] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [ ] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by PipoCanaja** [comment]
Hi @Calvario 
This fix is indeed necessary. It would be a lot better to also fix the discovery code (where the `discover_sensor` function is called) to ensure it is not called with `0` as `$divisor` parameter.

**Comment by PipoCanaja** [comment]
I would start with the sensors of CheckPoint : we need to fix any call with the divisor set to 0. If you run the discovery in "verbose" mode (`./discovery.php -v -d -h <device>`) you should find where the divisor is zero. 

The fix in [includes/discovery/sensors/entity-sensor.inc.php](https://github.com/librenms/librenms/pull/16464/files#diff-130444bfca104534cbbfa4c4fee55fd056c3790528b3f2c0a347a269d15a8b6b) would "mask" any issue and replace them by `$current = 0` which will make troubleshooting difficult if not impossible.

**Comment by Calvario** [comment]
I updated the issue, it look's like that it's related with how we handle entPhySensorScale & entPhySensorPrecision.

Indeed this fix would hide the issue but we could add a log or an eventlog to help find out about it.
I already used ./discovery.php in debug and verbose but the device as too many sensors, it make it tedious to check.
I just added (not committed yet) a log inside the "if" to find all the cases.
What do you think ? I think and eventlog is more suitable as we already have the "Error discovering sensors module. Check log file for more details." so it's easier for an user to link them.

As entPhySensorScale "zepto" is not handled, one way to fix it for the current issue would be to add another if like:
```
if ($entry['entPhySensorScale'] == 'zepto') {
    $divisor = '1000000000000000000';
    $multiplier = '1';
}
```

**Comment by murrant** [comment]
Improving the multiplier/scale detction would be much better...  Having it default to 1 instead of 0 is ideal :)

**Comment by Calvario** [comment]
Like this @murrant ?

**Comment by murrant** [comment]
Mostly yes, but I would assign [$divisor, $multiplier] from the match instead of setting it inside the match

**Comment by murrant** [comment]
oh, and the multiplier/divisor should be integers, not strings.  (old code set them as strings)

**Comment by Calvario** [comment]
Something like that then ?
```
// Try to handle the scale
[$divisor, $multiplier] = match ($entry['entPhySensorScale']) {
    'zepto' => [1000000000000000000, 1],
    'nano' => [1000000000, 1],
    'micro' => [1000000, 1],
    'milli' => [1000, 1],
    'units' => [1, 1],
    'kilo' => [1, 1000],
    'mega' => [1, 1000000],
    'giga' => [1, 1000000000],
    'yocto' => [1, 1],
    default => [1, 1],
};
```

File: includes/discovery/sensors/entity-sensor.inc.php [hunk 1]
```diff
@@ -112,41 +112,23 @@
...
+                'units' => [$divisor, $multiplier] = ['1', '1'],
+                'kilo' => [$divisor, $multiplier] = ['1', '1000'],
+                'mega' => [$divisor, $multiplier] = ['1', '1000000'],
+                'giga' => [$divisor, $multiplier] = ['1', '1000000000'],
+                'yocto' => [$divisor, $multiplier] = ['1', '1'],
+                default => [$divisor, $multiplier] = ['1', '1'],
+            };
+
+            if (is_numeric($entry['entPhySensorPrecision']) && $entry['entPhySensorPrecision'] > 0) {
+                $divisor .= str_pad('', $entry['entPhySensorPrecision'], '0');
```
Comment (murrant) [inline includes/discovery/sensors/entity-sensor.inc.php:131]: This code is actually duplicated in cisco-entity-sensors too :D

Comment (murrant) [reply includes/discovery/sensors/entity-sensor.inc.php:131]: Too many questions :)  Mostly, this code base has a lot of legacy.

**Review by murrant — APPROVED** [review]

**Pull request merged**